src/OpenTelemetry.Collector.Dependencies/Implementation/AzureSdkDiagnosticListener.cs

```diff
@@ -60,17 +60,18 @@ public override void OnStartActivity(Activity current, object valueValue)
                 }
             }
 
-            List<Link> links = null;
+            Func<IEnumerable<Link>> linksGetter = null;
             if (LinksPropertyFetcher.Fetch(valueValue) is IEnumerable<Activity> activityLinks)
             {
-                links = new List<Link>();
+                var links = new List<Link>();
+                linksGetter = () => links;
```

Comment:
This is allocation. We need to avoid it or allow IEnumerable overload

Reply:
I guess the best option is also the ugliest. But I get 1 less allocation then the current code:

<img width="793" alt="image" src="https://user-images.githubusercontent.com/1633368/66795775-a24d5d00-eed3-11e9-93e4-e450a6c15612.png">

Reply:
I can also just simplify to:

```csharp
Func<IEnumerable<Link>> linksGetter = null;
if (LinksPropertyFetcher.Fetch(valueValue) is IEnumerable<Activity> activityLinks)
{
    var links = new List<Link>();
    foreach (var link in activityLinks)
    {
        links.Add(new Link(new SpanContext(link.TraceId, link.ParentSpanId, link.ActivityTraceFlags)));
    }

    if (links.Count > 0)
    {
        linksGetter = () => links;
    }
}
```

But the list will be allocated even if there's nothing to enumerate to add to it.

Reply:
() => {} is an allocation. Can we avoid it even if there are links? There was no allocation here with current API.

Reply:
Right, the lambda will always allocate. 
The only option here is to add an overload to the whole API

Reply:
overload is fine.